### DISCUSSION TITLE
How to verify user's membership is still valid before accessing each page?

### DISCUSSION THREAD

**Discussion by Rongeegee** [vercel/next.js#53953] [Help]
### Summary

Hey guys, I need help. 

This problem really bugs me out, and I have no clue what to do. So, each user in the database has a field called 'membership_expiration_date'. If the membership is expired, user will be redirected to the payment page. I tried to use middleware to achieve this goal, however middleware does not allow database call as I will call edge function. I can verify the user's membership by checking if it's expired before each page load. However, this is very tedious and error-prone. Can anyone tell me any better way to do it?

**Comment by Nadeera3784** [comment]
Hi @Rongeegee ,
How about using a custom hook for this?
for example : 

```
const useMembership = () => {
  const [membership, setMembership] = useState(null);

  const fetchMembership = async () => {
    const response = await fetch(`/api/membership/${user.id}`);
    const data = await response.json();
    setMembership(data);
  };

  useEffect(() => {
    fetchMembership();
  }, []);

  return membership;
};

const App = () => {
  const membership = useMembership();

  if (!membership || membership.expirationDate < new Date()) {
    return <Redirect to="/payment" />;
  }

  return <Home />;
};
```

**Reply by Rongeegee** [reply]
Do I have to do this on every page? If so, that's very tedious and error-prone.

**Reply by Nadeera3784** [reply]
@Rongeegee ,
No don't have to do this on every page. You can use a higher-order component (HOC) to wrap all of your pages and check the user's membership before rendering them. for example :

```
const withMembershipCheck = (Component) => {
  return (props: Props) => {
    const membership = useMembership();

    if (!membership || membership.expirationDate < new Date()) {
      return <Redirect to="/payment" />;
    }

    return <Component {...props} />;
  };
};
 
```

and then You can use this HOC to wrap all of your pages, like so:

```
const Home = () => {
  return (
    <Layout>
      <h1>Home</h1>
    </Layout>
  );
};

const withMembershipCheck = withMembershipCheck(Home);

export default withMembershipCheck;
```

**Reply by Rongeegee** [reply]
In your case, you pass the Home component as a argument to withMembershipCheck function. Don't you have to do it to every page component ?

**Comment by devjiwonchoi** [comment]
I'd suggest saving `membership_expiration_date` info in a cookie on initial validation(auth like login, signup).

The `middleware.ts` provides cookie validation out of the box.

```ts
// middleware.ts
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

export function middleware(request: NextRequest) {
  const hasExpDate = request.cookies.has('membership_expiration_date')

  const response = NextResponse.next() // intiate response
  if (!hasExpDate) {
    // query DB for user's membership expiration date
    const expDate = '2023-12-31' // dummy date
    response.cookies.set('membership_expiration_date', expDate)
  }

  return response
}
```

### Reference

- [Using Cookies](https://nextjs.org/docs/app/building-your-application/routing/middleware#using-cookies)
